### PR TITLE
HHH-1152 Discriminator based inheritance for embeddable types

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-1152

### CODE REVIEW

**Comment by gavinking** [comment]
Great to have this finally, after all these years :)

File: hibernate-core/src/main/java/org/hibernate/mapping/Component.java [hunk 1]
```diff
@@ -73,8 +75,12 @@ public class Component extends SimpleValue implements MetaAttributable, Sortable
 	private boolean isKey;
 	private Boolean isGeneric;
 	private String roleName;
+	private Value discriminator;
+	private Map<Object, String> discriminatorValues;
+	private Map<String, String> subclassToSuperclass;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/mapping/Component.java:80]: Seems you're not copying this in the clone constructor.

File: hibernate-core/src/main/java/org/hibernate/mapping/Component.java [hunk 2]
```diff
@@ -172,9 +199,13 @@ public void addColumn(Column column) {
 	@Override
 	public List<Selectable> getSelectables() {
 		if ( cachedSelectables == null ) {
-			cachedSelectables = properties.stream()
+			final List<Selectable> selectables = properties.stream()
 					.flatMap( p -> p.getSelectables().stream() )
 					.collect( toList() );
+			if ( discriminator != null ) {
+				selectables.addAll( discriminator.getSelectables() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/mapping/Component.java:206]: Adding the discriminator to the selectables definitevly makes things easier, though I would like to discuss the implications of that. For entities, we do not treat the discriminator to be part of the entity state, because that is encoded in the entity key through the entity name.
I understand that embeddable data does not have such a key where the type/discriminator is stored, so this might very well be the only option to model this. Let's have a chat about it anyway :)

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java [hunk 3]
```diff
@@ -36,37 +42,65 @@
...
-		super(
-				bootDescriptor,
-				resolveEmbeddableJavaType( bootDescriptor, compositeUserType, creationContext ),
-				creationContext
-		);
+		this.embeddableJavaType = resolveEmbeddableJavaType( bootDescriptor, compositeUserType, creationContext );
 
+		final int propertySpan = bootDescriptor.getPropertySpan();
+		this.propertyAccesses = new PropertyAccess[propertySpan];
+		this.attributeNameToPositionMap = new ConcurrentHashMap<>( propertySpan );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:69]: I think a regular hash map should be fine

```suggestion
		this.attributeNameToPositionMap = new HashMap<>( propertySpan );
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java [hunk 4]
```diff
@@ -83,39 +117,75 @@ private static <T> JavaType<T> resolveEmbeddableJavaType(
...
+		return new EmbeddableInstantiatorPojoStandard( embeddableClass, runtimeDescriptorAccess );
 	}
 
-	@Override
-	protected PropertyAccess buildPropertyAccess(Property bootAttributeDescriptor) {
+	private PropertyAccess buildPropertyAccess(
+			Property bootAttributeDescriptor,
+			Class<?> embeddableClass,
+			boolean requireSetters) {
 		PropertyAccessStrategy strategy = bootAttributeDescriptor.getPropertyAccessStrategy( getEmbeddableJavaType().getJavaTypeClass() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:189]: Maybe this is wrong?

```suggestion
		PropertyAccessStrategy strategy = bootAttributeDescriptor.getPropertyAccessStrategy( embeddableClass );
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java [hunk 5]
```diff
@@ -187,13 +259,71 @@ private ReflectionOptimizer buildReflectionOptimizer(
...
+	}
+
 	@Override
 	public RepresentationMode getMode() {
 		return RepresentationMode.POJO;
 	}
 
 	@Override
 	public EmbeddableInstantiator getInstantiator() {
+		return getInstantiatorForSubclass( getEmbeddableJavaType().getJavaTypeClass().getName() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:315]:
```suggestion
		assert instantiators.size() == 1;
		return instantiators.iterator().values().iterator().next();
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/EmbeddableDiscriminatorConverter.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+	public EmbeddableDiscriminatorConverter(
+			NavigableRole discriminatorRole,
+			JavaType<O> domainJavaType,
+			JavaType<R> relationalJavaType,
+			List<DiscriminatorValueDetails> valueMappings) {
+		super( discriminatorRole, domainJavaType, relationalJavaType );
+
+		this.discriminatorValueToDetailsMap = CollectionHelper.concurrentMap( valueMappings.size() );
+		this.embeddableClassNameToDetailsMap = CollectionHelper.concurrentMap( valueMappings.size() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/EmbeddableDiscriminatorConverter.java:59]: Regular maps should be fine

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java (hunk 3 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:69]: This was copied from `AbstractEmbeddableRepresentationStrategy`, which I merged into EmbeddableRepresentationStrategyPojo` since it was the only implementor and it simplified the code. I will try a normal `HashMap` if you think it's better.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 7]
```diff
@@ -693,6 +811,58 @@ public <T> DomainResult<T> createDomainResult(
...
+		}
+		return null;
+	}
+
+	@Override
+	protected Object[] getAttributeValues(Object compositeInstance) {
+		final Object[] results = new Object[getNumberOfAttributeMappings()];
+		for ( int i = 0; i < getNumberOfAttributeMappings(); i++ ) {
+			final AttributeMapping attributeMapping = getAttributeMapping( i );
+			if ( declaresAttribute( compositeInstance.getClass().getName(), attributeMapping ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:837]:
```suggestion
		final String compositeClassName = compositeInstance.getClass().getName();
		for ( int i = 0; i < getNumberOfAttributeMappings(); i++ ) {
			final AttributeMapping attributeMapping = getAttributeMapping( i );
			if ( declaresAttribute( compositeClassName, attributeMapping ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/mapping/Component.java (hunk 2 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/mapping/Component.java:206]: IIRC this might have only been needed to pass validation against the component's columns. Maybe I'm wrong, but if that's the case we might treat it specially there.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 8]
```diff
@@ -693,6 +811,58 @@ public <T> DomainResult<T> createDomainResult(
...
+			}
+		}
+		return results;
+	}
+
+	@Override
+	protected void setAttributeValues(Object component, Object[] values) {
+		for ( int i = 0; i < values.length; i++ ) {
+			final AttributeMapping attributeMapping = getAttributeMapping( i );
+			if ( declaresAttribute( component.getClass().getName(), attributeMapping ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:854]:
```suggestion
		final String compositeClassName = component.getClass().getName();
		for ( int i = 0; i < values.length; i++ ) {
			final AttributeMapping attributeMapping = getAttributeMapping( i );
			if ( declaresAttribute( compositeClassName, attributeMapping ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 9]
```diff
@@ -721,12 +891,15 @@ public <X, Y> int breakDownJdbcValues(
 					);
 				}
 			}
+			if ( isPolymorphic() ) {
+				span += discriminatorMapping.breakDownJdbcValues( values[i], offset + span, x, y, valueConsumer, session );
+			}
 		}
 		else {
 			for ( int i = 0; i < size; i++ ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:899]:
```suggestion
			final String compositeClassName = domainValue.getClass().getName();
			for ( int i = 0; i < size; i++ ) {
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 10]
```diff
@@ -721,12 +891,15 @@ public <X, Y> int breakDownJdbcValues(
...
 				final AttributeMapping attributeMapping = attributeMappings.get( i );
 				if ( !attributeMapping.isPluralAttributeMapping() ) {
-					final Object attributeValue = domainValue == null
+					final Object attributeValue = domainValue == null || !declaresAttribute( domainValue.getClass().getName(), attributeMapping )
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:902]:
```suggestion
					final Object attributeValue = domainValue == null || !declaresAttribute( compositeClassName, attributeMapping )
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 11]
```diff
@@ -739,6 +912,10 @@ public <X, Y> int breakDownJdbcValues(
 					);
 				}
 			}
+			if ( isPolymorphic() ) {
+				final Object d = domainValue == null ? null : discriminatorMapping.getDiscriminatorValue( domainValue.getClass().getName() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:916]:
```suggestion
				final Object d = domainValue == null ? null : discriminatorMapping.getDiscriminatorValue( compositeClassName );
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 12]
```diff
@@ -755,27 +932,35 @@ public <X, Y> int decompose(
...
 				final Object attributeValue = values[ i ];
 				span += attributeMapping.decompose( attributeValue, offset + span, x, y, valueConsumer, session );
 			}
+			if ( isPolymorphic() ) {
+				span += discriminatorMapping.decompose( values[i], offset + span, x, y, valueConsumer, session );
+			}
 		}
 		else {
-			for ( int i = 0; i < attributeMappings.size(); i++ ) {
+			for ( int i = 0; i < size; i++ ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:951]:
```suggestion
			final String compositeClassName = domainValue.getClass().getName();
			for ( int i = 0; i < size; i++ ) {
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 13]
```diff
@@ -755,27 +932,35 @@ public <X, Y> int decompose(
...
 				final AttributeMapping attributeMapping = attributeMappings.get( i );
-				if ( !(attributeMapping instanceof PluralAttributeMapping )) {
-					final Object attributeValue = domainValue == null
+				if ( !attributeMapping.isPluralAttributeMapping() ) {
+					final Object attributeValue = domainValue == null || !declaresAttribute( domainValue.getClass().getName(), attributeMapping )
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:954]:
```suggestion
					final Object attributeValue = domainValue == null || !declaresAttribute( compositeClassName, attributeMapping )
```

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java [hunk 14]
```diff
@@ -755,27 +932,35 @@ public <X, Y> int decompose(
...
 							? null
 							: attributeMapping.getPropertyAccess().getGetter().get( domainValue );
 					span += attributeMapping.decompose( attributeValue, offset + span, x, y, valueConsumer, session );
 				}
 			}
+			if ( isPolymorphic() ) {
+				final Object d = domainValue == null ? null : discriminatorMapping.getDiscriminatorValue( domainValue.getClass().getName() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java:961]:
```suggestion
				final Object d = domainValue == null ? null : discriminatorMapping.getDiscriminatorValue( compositeClassName );
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/metamodel/attributeInSuper/WorkOrder.java [hunk 15]
```diff
@@ -8,15 +8,16 @@
...
+import jakarta.persistence.Id;
 import jakarta.persistence.OneToMany;
 import jakarta.persistence.OrderBy;
 
 @jakarta.persistence.Entity
 public class WorkOrder implements Serializable {
-	@EmbeddedId
-	private WorkOrderId id;
+	@Id
+	private Long id;
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/jpa/metamodel/attributeInSuper/WorkOrder.java:20]: Why this change?

Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/jpa/metamodel/attributeInSuper/WorkOrder.java:20]: This test was weird, with a component annotated both `@Embeddable` and `@MappedSuperclass` that was extended by another `@Embeddable`. Before this was "ignored", though now that we have inheritance the semantics have changed.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/EmbeddableDiscriminatorConverter.java (hunk 6 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/EmbeddableDiscriminatorConverter.java:59]: I remember trying to use simple `HashMap`s and getting some test failures, but now it looks like they work. Maybe it was just a temporary workaround I needed since I used `Class` as keys early on.

File: hibernate-core/src/main/java/org/hibernate/type/ComponentType.java [hunk 16]
```diff
@@ -745,7 +751,7 @@ public Object extract(CallableStatement statement, String paramName, SharedSessi
 	}
 
 	private Object resolve(Object[] value, SharedSessionContractImplementor session) throws HibernateException {
-		return instantiator().instantiate( () -> value, session.getFactory() );
+		return instantiator( null ).instantiate( () -> value, session.getFactory() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:754]: Shouldn't you pass the component discriminator here to allow resolving the correct instantiator?

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:754]: I don't have access to the discriminator value because it's not included in the values array.

File: hibernate-core/src/test/java/org/hibernate/orm/test/inheritance/embeddable/EmbeddableInheritanceAssciationsTest.java [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+	public void testManyToOne(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			// insert
+			final AssociatedEntity associated = session.find( AssociatedEntity.class, 1L );
+			session.persist( new TestEntity( 1L, new AssociationChildOne( "embeddable_1", associated ) ) );
+		} );
+		scope.inTransaction( session -> {
+			// queries
+			final AssociatedEntity result = session.createQuery(
+					"select embeddable.manyToOne from TestEntity where id = 1",
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/inheritance/embeddable/EmbeddableInheritanceAssciationsTest.java:56]: Yes, duplicate columns error.

File: release-announcement.adoc [hunk 18]
```diff
@@ -98,3 +98,52 @@ class MyEmbeddable { ... }
...
+
+When retrieving the inherited embedded property, Hibernate will read the discriminator value and instantiate the
+correct `@Embeddable` subtype with its corresponding properties.
+
+For example, a mapping like this:
+[source,java]
+----
+@Embeddable
+@DiscriminatorValue( "parent" )
+@DiscriminatorColumn( name = "embeddable_type" )
```
Comment (beikov) [inline release-announcement.adoc:116]: I guess `@DiscriminatorFormula` is also supported? Maybe that's worth mentioning?

Comment (mbladel) [reply release-announcement.adoc:116]: Sure, I'll mention it in the user guide.

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java [hunk 19]
```diff
@@ -187,13 +259,71 @@ private ReflectionOptimizer buildReflectionOptimizer(
...
+	}
+
+	@Override
+	public EmbeddableInstantiator getInstantiatorForSubclass(String embeddableClassName) {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:319]: What's the difference? The initializer has easy access to the discriminator mapping to then retrieve the embeddable class name, while the representation strategy does not (unless we want to create a different map with discriminator values there). Also this is analogous to the way entity initializers determine the concrete persister to use.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java [hunk 20]
```diff
@@ -451,6 +611,75 @@ private static List<PropertyData> collectClassElements(
...
 
+	private static void collectSubclassElements(
+			AccessType propertyAccessor,
+			MetadataBuildingContext context,
+			XClass superclass,
+			List<PropertyData> classElements,
+			BasicType<?> discriminatorType,
+			Map<Object, String> discriminatorValues,
+			Map<String, String> subclassToSuperclass) {
+		for ( final XClass subclass : context.getMetadataCollector().getEmbeddableSubclasses( superclass ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java:622]: Is this "safe" i.e. are all embeddable subclasses known at this point for sure?

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java:622]: Yes, I believe so, all embeddable subclasses are registered in `org.hibernate.boot.model.internal.AnnotationBinder#buildInheritanceStates`

File: hibernate-core/src/test/java/org/hibernate/orm/test/inheritance/embeddable/EmbeddableInheritanceAssciationsTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/test/java/org/hibernate/orm/test/inheritance/embeddable/EmbeddableInheritanceAssciationsTest.java:56]: As we discussed, this should be fixed i.e. column reuse in different subtypes should be allowed.

File: hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java [hunk 21]
```diff
@@ -115,20 +118,42 @@ public static Object[] getJdbcValues(
...
+		assert jdbcIndex == valueCount;
 		return jdbcValues;
 	}
 
+	public static Object[] getValues(EmbeddableMappingType embeddableMappingType, Object domainValue) {
+		final Object[] attributeValues = embeddableMappingType.getValues( domainValue );
+		if ( embeddableMappingType.isPolymorphic() ) {
+			final Object[] result = new Object[attributeValues.length + 1];
+			System.arraycopy( attributeValues, 0, result, 0, attributeValues.length );
+			result[attributeValues.length] = domainValue.getClass().getName();
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java:138]: No, when serializing struct values the discriminator converter will be called transforming the class name to the correct relational value.

File: hibernate-core/src/main/java/org/hibernate/type/ComponentType.java (hunk 16 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/type/ComponentType.java:754]: As we discussed, this method should get the discriminator value passed.

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:319]: >Also this is analogous to the way entity initializers determine the concrete persister to use.

Sounds like something we should improve on :)

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java:319]: See https://hibernate.atlassian.net/browse/HHH-18104
